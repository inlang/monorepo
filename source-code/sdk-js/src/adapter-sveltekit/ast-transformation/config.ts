import { loadFile, type ProxifiedModule } from "magicast"
import { writeFile } from "node:fs/promises"
import { initConfig } from '../../config/index.js'
import { stat } from "node:fs/promises"

export const doesPathExist = async (path: string) => !!(await stat(path).catch(() => undefined))

export type TransformConfig = {
	languageInUrl: boolean
	isStatic: boolean
	srcFolder: string
	rootRoutesFolder: string
	hasAlreadyBeenInitialized: boolean
	sourceFileName?: string
	sourceMapName?: string
	isTypeScriptProject: boolean
}

const cwd = process.cwd()

let configPromise: Promise<TransformConfig> | undefined = undefined

export const getTransformConfig = async (): Promise<TransformConfig> => {
	if (configPromise) return configPromise

	// eslint-disable-next-line no-async-promise-executor
	return configPromise = new Promise<TransformConfig>(async (resolve) => {
		const srcFolder = cwd + "/src"
		const routesFolder = srcFolder + "/routes"

		await createInlangConfigIfNotPresentYet()

		const inlangConfig = await initConfig()

		const languageInUrl = inlangConfig?.sdk?.languageNegotiation?.strategies?.some(({ type }) => type === 'url') || false

		const rootRoutesFolder = routesFolder + "/(app)" + (languageInUrl ? "/[lang]" : "")
		const isStatic = await shouldContentBePrerendered(rootRoutesFolder)

		const hasAlreadyBeenInitialized = await doesPathExist(rootRoutesFolder)

		const isTypeScriptProject = await doesPathExist(cwd + '/tsconfig.json')

		resolve({
			languageInUrl,
			isStatic,
			srcFolder,
			rootRoutesFolder,
			hasAlreadyBeenInitialized,
			isTypeScriptProject,
		})
	})
}

export const resetConfig = () => configPromise = undefined

// ------------------------------------------------------------------------------------------------

const createInlangConfigIfNotPresentYet = async () => {
	const inlangConfigPath = cwd + "/inlang.config.js"
	const inlangConfigExists = await doesPathExist(inlangConfigPath)
	if (inlangConfigExists) return

	// TODO: use sdkPlugin once ready https://github.com/inlang/inlang/issues/610

	return writeFile(inlangConfigPath, `
/**
 * @type { import("@inlang/core/config").DefineConfig }
 */
export async function defineConfig(env) {
	const { default: jsonPlugin } = await env.$import(
		"https://cdn.jsdelivr.net/gh/samuelstroschein/inlang-plugin-json@2/dist/index.js"
	)
	const { default: sdkPlugin } = await env.$import(
		"https://cdn.jsdelivr.net/npm/@inlang/sdk-js/dist/plugin/index.js"
	)

	return {
		referenceLanguage: "en",
		plugins: [
			jsonPlugin({ pathPattern: "./languages/{language}.json" }),
			sdkPlugin({
				languageNegotiation: {
					strategies: [{ type: "localStorage" }, { type: "navigator" }]
				}
			}),
		],
	}
}
`)
}

// ------------------------------------------------------------------------------------------------

const shouldContentBePrerendered = async (routesFolder: string) => {
	const filesToLookFor = [
		'+layout.server.js',
		'+layout.server.ts',
		'+layout.js',
		'+layout.ts',
	]

	const modules = (await Promise.all(
		filesToLookFor.map(file => loadFile(routesFolder + `/${file}`).catch(() => undefined))
	))
		.filter(Boolean) as ProxifiedModule<any>[]

	return modules.map(mod => [true, 'auto'].includes(mod.exports.prerender))
		.some(Boolean)
}