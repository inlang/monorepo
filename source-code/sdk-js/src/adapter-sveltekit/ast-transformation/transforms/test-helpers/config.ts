import type { TransformConfig } from "../../config.js"

export const baseTestConfig: TransformConfig = {
	isStatic: false,
	languageInUrl: false,
	srcFolder: "",
	rootRoutesFolder: "",
	hasAlreadyBeenInitialized: false,
	sourceFileName: "",
	sourceMapName: "",
	isTypeScriptProject: false
}
