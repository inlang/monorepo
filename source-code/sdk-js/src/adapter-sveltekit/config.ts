import { stat } from "node:fs/promises"

export const doesPathExist = async (path: string) => !!(await stat(path).catch(() => undefined))

export const getConfig = async () => {
	// TODO: read information from files
	const isSPA = false
	const isStatic = false

	const srcFolder = process.cwd() + '/src'
	const rootRoutesFolder = srcFolder + '/routes/(app)' + (isSPA ? '' : '/[lang]')

	const hasAlreadyBeenInitialized = await doesPathExist(rootRoutesFolder)

	return {
		isSPA,
		isStatic,
		srcFolder,
		rootRoutesFolder,
		hasAlreadyBeenInitialized,
	}
}