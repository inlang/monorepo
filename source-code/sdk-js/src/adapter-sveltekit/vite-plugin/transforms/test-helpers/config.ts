import type { TransformConfig } from "../../config.js"
import { deepmerge } from "deepmerge-ts"

type DeepPartial<T> = T extends Record<PropertyKey, unknown>
	? {
			[Key in keyof T]?: DeepPartial<T[Key]>
	  }
	: T

export const getTransformConfig = (overrides: DeepPartial<TransformConfig> = {}): TransformConfig =>
	deepmerge(
		{
			isStatic: false,
			languageInUrl: false,
			cwdFolderPath: "",
			rootRoutesFolder: "",
			sourceFileName: "",
			sourceMapName: "",
			inlang: {
				referenceLanguage: "en",
				languages: ["en", "de"],
				readResources: async () => [],
				writeResources: async () => undefined,
				sdk: {
					debug: false,
					languageNegotiation: {
						strict: false,
						strategies: [{ type: "localStorage", key: "language" }],
					},
					resources: {
						cache: "build-time",
					},
				},
			},
			svelteKit: {
				usesTypeScript: false,
				version: undefined,
				files: {
					appTemplate: "src/app.html",
					routes: "src/routes",
					serverHooks: "src/hooks.server",
				},
			},
		} satisfies TransformConfig,
		overrides as any,
	)
