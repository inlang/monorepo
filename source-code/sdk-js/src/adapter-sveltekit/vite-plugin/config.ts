import { loadFile, type ProxifiedModule } from "magicast"
import { mkdir, readFile, writeFile, stat } from "node:fs/promises"
import { initConfig } from '../../config/index.js'
import { dedent } from 'ts-dedent'
import type { InlangConfig } from '@inlang/core/config'
import { testConfigFile } from '@inlang/core/test'
import { initInlangEnvironment, InlangConfigWithSdkProps } from '../../config/config.js'
import { validateSdkConfig } from '@inlang/sdk-js-plugin'
// @ts-ignore
import pkg from '../../../package.json'

export const doesPathExist = async (path: string) => !!(await stat(path).catch(() => false))

export type TransformConfig = {
	debug?: boolean
	languageInUrl: boolean
	isStatic: boolean
	srcFolder: string
	rootRoutesFolder: string
	sourceFileName?: string
	sourceMapName?: string
	isTypeScriptProject: boolean
}

const cwd = process.cwd()

let configPromise: Promise<TransformConfig> | undefined = undefined

export const getTransformConfig = async (): Promise<TransformConfig> => {
	if (configPromise) return configPromise

	// eslint-disable-next-line no-async-promise-executor
	return configPromise = new Promise<TransformConfig>(async (resolve) => {
		const srcFolder = cwd + "/src"
		const routesFolder = srcFolder + "/routes"

		await createInlangConfigIfNotPresentYet()
		await updateSdkPluginVersion()

		// TODO: combine `testConfigFile` and `initConfig` functionality
		const inlangConfigAsString = await readFile(cwd + "/inlang.config.js", { encoding: "utf-8" })
		const [, exception] = await testConfigFile({
			file: inlangConfigAsString,
			env: await initInlangEnvironment()
		})
		if (exception) {
			throw exception
		}

		const inlangConfigModule = await import(cwd + "/inlang.config.js")
		const inlangConfig = await initConfig(inlangConfigModule)

		assertConfigWithSdk(inlangConfig)
		inlangConfig.sdk = validateSdkConfig(inlangConfig.sdk)

		const languageInUrl = inlangConfig?.sdk?.languageNegotiation?.strategies?.some(({ type }) => type === 'url') || false

		const rootRoutesFolder = routesFolder + "/" + (languageInUrl ? "[lang]" : "")
		const isStatic = await shouldContentBePrerendered(routesFolder) || await shouldContentBePrerendered(rootRoutesFolder)

		const isTypeScriptProject = await doesPathExist(cwd + '/tsconfig.json')

		resolve({
			debug: !!inlangConfig.sdk?.debug,
			languageInUrl,
			isStatic,
			srcFolder,
			rootRoutesFolder,
			isTypeScriptProject,
		})
	})
}

export const resetConfig = () => configPromise = undefined

// ------------------------------------------------------------------------------------------------

class InlangSdkConfigError extends Error { }

function assertConfigWithSdk(config: InlangConfig | undefined): asserts config is InlangConfigWithSdkProps {
	if (!config) {
		throw new InlangSdkConfigError('Could not find `inlang.config.js` in the root of your project.`')
	}

	if (!('sdk' in config)) {
		// TODO: link to docs
		throw new InlangSdkConfigError('The `sdk` property is missing in your `inlang.config.js` file.`. Make sure to use the `sdkPlugin` in your `inlang.config.js` file.')
	}
}

// ------------------------------------------------------------------------------------------------

const createInlangConfigIfNotPresentYet = async () => {
	const inlangConfigPath = cwd + "/inlang.config.js"
	const inlangConfigExists = await doesPathExist(inlangConfigPath)
	if (inlangConfigExists) return

	await createDemoResources()

	return writeFile(inlangConfigPath, `
/**
 * @type { import("@inlang/core/config").DefineConfig }
 */
export async function defineConfig(env) {
	const { default: jsonPlugin } = await env.$import(
		"https://cdn.jsdelivr.net/gh/samuelstroschein/inlang-plugin-json@2/dist/index.js"
	)
	const { default: sdkPlugin } = await env.$import(
		"https://cdn.jsdelivr.net/npm/@inlang/sdk-js-plugin/dist/index.js"
	)

	return {
		referenceLanguage: "en",
		plugins: [
			jsonPlugin({
				pathPattern: "./languages/{language}.json",
			}),
			sdkPlugin({
				languageNegotiation: {
					strategies: [{ type: "localStorage" }]
				}
			}),
		],
	}
}
`)
}

// TODO: do this in a better way
const createDemoResources = async () => {
	if (!await doesPathExist(cwd + '/languages')) {
		await mkdir(cwd + '/languages')
	}

	await writeFile(cwd + '/languages/en.json', dedent`
		{
		  "welcome": "Welcome to inlang"
		}
	`, { encoding: 'utf-8' })

	await writeFile(cwd + '/languages/de.json', dedent`
		{
		  "welcome": "Willkommen bei inlang"
		}
	`, { encoding: 'utf-8' })
}

// ------------------------------------------------------------------------------------------------

const shouldContentBePrerendered = async (routesFolder: string) => {
	const filesToLookFor = [
		'+layout.server.js',
		'+layout.server.ts',
		'+layout.js',
		'+layout.ts',
	]

	const modules = (await Promise.all(
		filesToLookFor.map(file => loadFile(routesFolder + `/${file}`).catch(() => undefined))
	))
		.filter(Boolean) as ProxifiedModule<any>[]

	return modules.map(mod => [true, 'auto'].includes(mod.exports.prerender))
		.some(Boolean)
}

// ------------------------------------------------------------------------------------------------

const updateSdkPluginVersion = async () => {
	const inlangConfigAsString = await readFile(cwd + "/inlang.config.js", { encoding: "utf-8" })
	const version = pkg.version

	// this regex detects the new `https://cdn.jsdelivr.net/npm/@inlang/sdk-js-plugin/dist/index.js` as well as
	// the older https://cdn.jsdelivr.net/npm/@inlang/sdk-js/dist/plugin/index.js url
	// both urls are also detected with the optional @version identifier
	const REGEX_PLUGIN_VERSION = /https:\/\/cdn\.jsdelivr\.net\/npm\/@inlang\/sdk-js(-plugin)?@?(.*)?\/dist(\/plugin)?\/index\.js/g

	const newConfig = inlangConfigAsString.replace(
		REGEX_PLUGIN_VERSION,
		`https://cdn.jsdelivr.net/npm/@inlang/sdk-js-plugin@${version}/dist/index.js`,
	)

	if (inlangConfigAsString !== newConfig) {
		console.info(`Updating 'inlang.config.js' to use the correct version of '@inlang/sdk-js-plugin' (${version})`)

		await writeFile(cwd + "/inlang.config.js", newConfig)
	}
}
