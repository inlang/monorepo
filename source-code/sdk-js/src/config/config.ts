import { InlangConfig, InlangConfigModule, setupConfig } from "@inlang/core/config"
import { initialize$import } from '@inlang/core/environment'
import fs from "node:fs/promises"
import { resolve } from "node:path"
import { SdkConfigInput, validateSdkConfig } from '../plugin/schema.js'

export type InlangConfigWithSdkProps = InlangConfig & {
	sdk?: SdkConfigInput
}

class InlangSdkConfigError extends Error { }

class InlangError extends Error { }

const readInlangConfig = async () => {
	const env = {
		$fs: fs,
		$import: initialize$import({
			fs,
			fetch: async (...args) => await fetch(...args)
				.catch(error => {
					// TODO: create an issue
					if (error instanceof TypeError && (error.cause as any)?.code === 'UND_ERR_CONNECT_TIMEOUT') {
						throw new InlangError(`

Node.js failed to resolve the URL. This can happen sometimes during development. Usually restarting the server helps.

	`, { cause: error })
					}

					throw error
				})
		})
	}
	const module = (await import(
		/* @vite-ignore */ resolve(process.cwd(), "./inlang.config.js")
	)) as InlangConfigModule

	return setupConfig({ module, env })
}

function assertConfigWithSdk(config: InlangConfig | undefined): asserts config is InlangConfigWithSdkProps {
	if (!config) {
		throw new InlangSdkConfigError('Could not find `inlang.config.js` in the root of your project.`')
	}

	if (!('sdk' in config)) {
		// TODO: link to docs
		throw new InlangSdkConfigError('The `sdk` property is missing in your `inlang.config.js` file.`. Make sure to use the `sdkPlugin` in your `inlang.config.js` file.')
	}
}

export const initConfig = async () => {
	const config = await readInlangConfig()
	assertConfigWithSdk(config)

	validateSdkConfig(config.sdk)

	return config
}
