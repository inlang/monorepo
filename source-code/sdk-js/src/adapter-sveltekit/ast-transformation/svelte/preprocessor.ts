import { dirname } from "node:path"
import { getConfig } from "../config.js"
import { transformCode } from './transforms.js'

type FileType = "+layout.svelte" | "+page.svelte" | "*.svelte"

export type FileInformation = {
	type: FileType
	root: boolean
}

const getFileInformation = (srcFolder: string, id: string): FileInformation | undefined => {
	if (!id.startsWith(srcFolder)) return undefined

	const path = id.replace(srcFolder, "")

	const dir = dirname(path)
	const root = dir.endsWith('/routes')

	if (path.endsWith("/+layout.svelte")) {
		return {
			type: "+layout.svelte",
			root,
		}
	}

	if (path.endsWith("/+page.svelte")) {
		return {
			type: "+page.svelte",
			root,
		}
	}

	if (path.endsWith(".svelte")) {
		return {
			type: "*.svelte",
			root: false,
		}
	}

	return undefined
}

// ------------------------------------------------------------------------------------------------

type PreprocessMarkupArgs = {
	content: string
	filename: string
}

export const preprocess = () => ({
	markup: async ({ content, filename }: PreprocessMarkupArgs) => {
		const config = await getConfig()

		const fileInformation = getFileInformation(config.srcFolder, filename)
		// eslint-disable-next-line unicorn/no-null
		if (!fileInformation) return null

		return { code: transformCode(config, content, fileInformation) }
	},
})
