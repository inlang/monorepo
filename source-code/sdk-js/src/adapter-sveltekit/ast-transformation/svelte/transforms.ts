
import type { TransformConfig } from '../config.js'
import type { FileInformation } from './preprocessor.js'
import { transformSvelte } from './transforms/*.svelte.js'
import { transformLayoutSvelte } from './transforms/+layout.svelte.js'
import { transformPageSvelte } from './transforms/+page.svelte.js'

// ------------------------------------------------------------------------------------------------

export const transformCode = (config: TransformConfig, code: string, { type, root }: FileInformation) => {
	switch (type) {
		case "+layout.svelte":
			return transformLayoutSvelte(config, code, root)
		case "+page.svelte":
			return transformPageSvelte(config, code, root)
		case "*.svelte":
			return transformSvelte(config, code)
	}
}
