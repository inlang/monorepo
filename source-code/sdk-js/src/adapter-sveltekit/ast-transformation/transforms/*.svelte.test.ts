import { describe, it } from "vitest"
import { transformSvelte } from "./*.svelte.js"
import { baseTestConfig } from "./test-helpers/config.js"
import type { TransformConfig } from "../config.js"
import { readFileSync } from "node:fs"

const testSvelteFile = readFileSync(__dirname + "/test-helpers/test.svelte").toString()
describe("transformSvelte", () => {
	it("languageInUrl is true", ({ expect }) => {
		const config: TransformConfig = {
			...baseTestConfig,
			languageInUrl: true,
		}
		const code = transformSvelte(config, testSvelteFile)
		expect(code).toMatchSnapshot()
	})

	it("languageInUrl is false", ({ expect }) => {
		const code = transformSvelte(baseTestConfig, testSvelteFile)
		expect(code).toMatchSnapshot()
	})
})
