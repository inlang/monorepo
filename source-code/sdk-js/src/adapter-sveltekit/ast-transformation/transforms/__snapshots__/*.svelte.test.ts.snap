// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformSvelte > languageInUrl is false 1`] = `
"<script lang=\\"ts\\">
    export let prop: string
</script>

<style>
    .red {
        color: red;
    }
</style>

<h1 class=\\"red\\">{prop}</h1>
"
`;

exports[`transformSvelte > languageInUrl is true 1`] = `
"<script lang=\\"ts\\">
    export let prop: string
</script>

<style>
    .red {
        color: red;
    }
</style>

<h1 class=\\"red\\">{prop}</h1>
"
`;
